File: android/app/src/main/java/com/example/mycarnavi/RoutesApiClient.kt
package com.example.mycarnavi

import android.util.Log
import com.google.android.gms.maps.model.LatLng
import com.google.maps.android.PolyUtil
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject

data class RouteInfo(
    val points: List<LatLng>,
    val distanceMeters: Int,
    val durationSeconds: Long
)

class RoutesApiClient(private val apiKey: String) {

    private val client = OkHttpClient()

    /** ブロッキング呼び出しのため、IOスレッドから呼ぶこと */
    fun fetchRoute(
        originLatitude: Double,
        originLongitude: Double,
        destinationLatitude: Double,
        destinationLongitude: Double
    ): RouteInfo? {

        val requestJson = JSONObject().apply {
            put("origin", waypoint(originLatitude, originLongitude))
            put("destination", waypoint(destinationLatitude, destinationLongitude))
            put("travelMode", "DRIVE")
            // 現在の交通状況（渋滞）を考慮した所要時間を返す
            put("routingPreference", "TRAFFIC_AWARE")
        }

        val request = Request.Builder()
            .url("https://routes.googleapis.com/directions/v2:computeRoutes")
            .addHeader("X-Goog-Api-Key", apiKey)
            .addHeader(
                "X-Goog-FieldMask",
                "routes.polyline.encodedPolyline,routes.distanceMeters,routes.duration"
            )
            .post(
                requestJson.toString()
                    .toRequestBody("application/json".toMediaType())
            )
            .build()

        client.newCall(request).execute().use { response ->

            val body = response.body?.string()

            if (!response.isSuccessful || body == null) {
                Log.e(TAG, "Routes API error: code=${response.code} body=$body")
                return null
            }

            val routes = JSONObject(body).optJSONArray("routes")

            if (routes == null || routes.length() == 0) {
                Log.w(TAG, "Routes API returned no routes: $body")
                return null
            }

            val route = routes.getJSONObject(0)

            val encodedPolyline = route
                .getJSONObject("polyline")
                .getString("encodedPolyline")

            return RouteInfo(
                points = PolyUtil.decode(encodedPolyline),
                distanceMeters = route.optInt("distanceMeters", 0),
                durationSeconds = route.optString("duration", "0s")
                    .removeSuffix("s")
                    .toLongOrNull() ?: 0L
            )
        }
    }

    private fun waypoint(latitude: Double, longitude: Double): JSONObject {
        return JSONObject().put(
            "location",
            JSONObject().put(
                "latLng",
                JSONObject()
                    .put("latitude", latitude)
                    .put("longitude", longitude)
            )
        )
    }

    companion object {
        private const val TAG = "RoutesApiClient"
    }
}

File: ios/Navi/Services/RoutesApiClient.swift
import CoreLocation
import Foundation
import OSLog

struct RoutesApiClient {
    private let apiKey: String
    private let logger = Logger(subsystem: "suzukiken.Navi", category: "RoutesApiClient")

    init(apiKey: String = MapsApiConfig.apiKey) {
        self.apiKey = apiKey
    }

    func fetchRoute(
        originLatitude: Double,
        originLongitude: Double,
        destinationLatitude: Double,
        destinationLongitude: Double
    ) async -> RouteInfo? {
        guard !apiKey.isEmpty else {
            logger.error("MAPS_API_KEY is not configured")
            return nil
        }

        let body: [String: Any] = [
            "origin": waypoint(latitude: originLatitude, longitude: originLongitude),
            "destination": waypoint(latitude: destinationLatitude, longitude: destinationLongitude),
            "travelMode": "DRIVE",
            "routingPreference": "TRAFFIC_AWARE",
        ]

        guard let url = URL(string: "https://routes.googleapis.com/directions/v2:computeRoutes"),
              let jsonData = try? JSONSerialization.data(withJSONObject: body)
        else {
            return nil
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.httpBody = jsonData
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "X-Goog-Api-Key")
        request.setValue(
            "routes.polyline.encodedPolyline,routes.distanceMeters,routes.duration",
            forHTTPHeaderField: "X-Goog-FieldMask"
        )

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
                logger.error("Routes API error: \((response as? HTTPURLResponse)?.statusCode ?? -1)")
                return nil
            }

            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let routes = json["routes"] as? [[String: Any]],
                  let route = routes.first,
                  let polyline = route["polyline"] as? [String: Any],
                  let encoded = polyline["encodedPolyline"] as? String
            else {
                logger.warning("Routes API returned no routes")
                return nil
            }

            let durationRaw = route["duration"] as? String ?? "0s"
            let durationSeconds = Int64(durationRaw.dropLast()) ?? 0

            return RouteInfo(
                points: PolylineDecoder.decode(encoded),
                distanceMeters: route["distanceMeters"] as? Int ?? 0,
                durationSeconds: durationSeconds
            )
        } catch {
            logger.error("Routes API request failed: \(error.localizedDescription)")
            return nil
        }
    }

    private func waypoint(latitude: Double, longitude: Double) -> [String: Any] {
        [
            "location": [
                "latLng": [
                    "latitude": latitude,
                    "longitude": longitude,
                ],
            ],
        ]
    }
}
